content/blog/2021-07-13-hyperparam-tuning.md

```diff
@@ -0,0 +1,295 @@
+
+picture: 2021-07-13/tuning-hyperparams.png
+pictureComment: Tuning Hyperparameters with Reproducible Experiments
+author: milecia_mcgregor
+commentsUrl: https://discuss.dvc.org/t/utilizing-custom-git-references-in-dvc/727
+tags:
+  - MLOps
+  - Git
+  - Experiments
+  - Reproducibility
+---
+
+## Intro
+
+When you're starting to build a new machine learning model and you're deciding
+on the model architecture, there are a number of issues that arise. You have to
+monitor code changes you make, note any differences in the data you've used for
+training, and keep up with hyperparameter value updates.
+
+Being able to track all of these changes is important so that you can reproduce
+your experiments without wondering which changes gave you the best model. You
+can go back to any point in your experimenting process to see which changes gave
+you the best results.
+
+In this post, we're going to go through an example of hyperparameter tuning with
+reproducibility using DVC. You can add this to any existing project you're
+working on or start from a fresh project.
+
+## Background on Hyperparameters
+
+Before jumping straight into training and experiments, let's briefly go over
+some background on hyperparameters.
+[Hyperparameters](/doc/command-reference/params) are the values that define your
+model. This includes things like the number of layers in a neural network or the
+learning rate for gradient descent.
+
+These parameters are different from model parameters because we can't get them
+from training our model. They are used to _create_ the model we train with.
+Optimizing these values means running training steps for different kinds of
+models to see how accurate the results are. We can get the best model from
+iterating through different hyperparameters values and seeing how they effect
+our accuracy.
+
+That's why we do hyperparameter tuning. There are a couple common methods that
+we'll do some code examples with: grid search and random search.
+
+## Tuning with DVC
+
+Let's start by talking about DVC a bit because we'll be using it to add
+reproducibility to our tuning process. This is the tool we'll be using to track
+changes in our data, code, and hyperparamters. With DVC, we can add some
+automation to the tuning process and be able to find and restore any really good
+models that emerge.
+
+A few things DVC makes easier to do:
+
+- Letting you make changes without worrying about finding them later
+- Onboarding other engineers to a project
+- Sharing experiments with other engineers on different machines
+- Tracking experiments and results in your Git repo and compare based on
+  parameters from the command line
+- Versioning the entire pipeline (including data, code, and commands run)
```

Comment:
Up to you and other reviewers whether to keep these last two bullets. They still seem a little long and incosistent with the others, but I think it's fine either way.

Reply:
I decided to just take them out.